Extracting out the code to versionize docs from mark-new-version and fixing a bunch of issues

nikhiljindal opened the pull request:
Extracting out the relevant code to versionize our docs into a separate script.
Have also added a script to verify that running versionize-docs multiple times does not produce any diffs.
Apart from the refactoring, this PR fixes the following issues as well:
- Fixed HTML_PREVIEW_PREFIX regex, so that HEAD in htmlpreview links is now replaced appropriately.
- Updated the IS_VERSIONED tag to be enclosed in "BEGIN MUNGE" and "END MUNGE" tags, so that it is not stripped off by gendocs (Right now, gendocs strips off IS_VERSIONED and hence all our versioned kubectl docs contain the unversioned warnings)
- Fixed analytics munger to not delete line breaks which it should not. Also, added a relevant test case. (Fixes https://github.com/GoogleCloudPlatform/kubernetes/issues/11676)
- Changed the ordering of mungedocs, so that checkHeaderLines mungedoc is run after unversioned warning is added, so that it can add line breaks after unversioned warnings.

Original description:
Extracted the relevant code to a separate script.
This PR also includes a temporary commit produced by running versionize-docs.

nikhiljindal committed:
Extract out the versionize docs part in a separate script

nikhiljindal commented:
cc @lavalamp @zmerlynn @brendandburns

lavalamp commented:
The new script looks great-- but it doesn't work on your new preview links.

lavalamp commented:
Specifically links like this one should have gotten changed:

https://htmlpreview.github.io/?https://github.com/GoogleCloudPlatform/kubernetes/HEAD/docs/api-reference/definitions.html#_v1_node

zmerlynn commented:
You can make preview commits by pushing a separate personal branch - that would make this PR mergeable.

zmerlynn commented:
I'm now OOO until morning.. It looks reasonable but make sure mark_new_version didn't get broken in the process.

thockin commented:
Looks obvious, but the analytics change seems unrelated?

On Mon, Jul 20, 2015 at 5:38 PM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *passed* for commit 2a8e75f
> <https://github.com/GoogleCloudPlatform/kubernetes/commit/2a8e75f741317b75fb523020077adb3e0a76e875>
> .
>
>    - Build Log
>    <https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/2a8e75f741317b75fb523020077adb3e0a76e875/kubernetes-pull-build-test-e2e-gce/3086/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/2a8e75f741317b75fb523020077adb3e0a76e875/kubernetes-pull-build-test-e2e-gce/3086/_artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//3086>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11622#issuecomment-123110155>
> .
>

nikhiljindal commented:
The analytics mungedocs was deleting empty lines in front of '<!-- TAG IS_VERSIONED -->, which wasnt looking nice in the docs.
The change in that file is not to delete empty lines as per the comment I received here: https://github.com/nikhiljindal/kubernetes/commit/cd6e83410f67a9ee039ec5d4fb3315a861cb84a7#commitcomment-12263778

If required, I can move that change to a different PR.

thockin commented:
I'm fine with it.

nikhiljindal commented:
Updated code as per comments.
PTAL.
Will delete the second commit, when I get an LGTM

thockin commented:
I can't final review right now.  @lavalamp

lavalamp commented:
Looking.

lavalamp commented:
Question-- why do we run mungedocs on the stripped docs??

nikhiljindal commented:
Have added the suggested test case to cmd/mungedocs/analytics_test.go and have removed the second commit, so that this PR is now ready to be merged.

zmerlynn commented:
What's going on with this commit? Why is @k8s-bot constantly testing it?

nikhiljindal commented:
Sorry I think k8s-bot is rerunning the test, each time I update my commit.

zmerlynn commented:
Oh, so you ARE re-pushing over and over again. :)

nikhiljindal commented:
Is there a way to stop k8s-bot until I am ready?

nikhiljindal commented:
Have fixed all comments as discussed.
Running versionize-docs repeatedly does not produce any diffs now. Have also added a script to verify that.
Apart from the refactoring, this PR fixes the following issues as well:
- Fixed HTML_PREVIEW_PREFIX regex, so that HEAD in htmlpreview links is now replaced appropriately.
- Updated the IS_VERSIONED tag to be enclosed in "BEGIN MUNGE" and "END MUNGE" tags, so that it is not stripped off by gendocs (Right now, gendocs strips off IS_VERSIONED and hence all our versioned kubectl docs contain the unversioned warnings)
- Fixed analytics munger to not delete line breaks which it should not. Also, added a relevant test case.
- Changed the ordering of mungedocs, so that checkHeaderLines mungedoc is run after unversioned warning is added, so that it can add line breaks after unversioned warnings.

PTAL

lavalamp commented on cmd/mungedocs/analytics_test.go:
```diff
@@ -27,53 +27,50 @@ func TestAnalytics(t *testing.T) {
-[![Analytics](https://kubernetes-site.appspot.com/UA-36037335-10/GitHub/path/to/file-name.md?pixel)]()
-` + endMungeTag("GENERATED_ANALYTICS") + `
-`},
+		{"aoeu", "aoeu" + "\n" + "\n" + "\n" +
```
Can you insert a newline after the { so that the indentation will be consistent between the two strings? It's very hard to read right now.

nikhiljindal replied:
Done

lavalamp commented on hack/verify-versionize-docs.sh:
```diff
@@ -0,0 +1,104 @@
+
+# Run versionize-docs.
+echo "Running versionize-docs for the first time"
+${KUBE_ROOT}/build/versionize-docs.sh v1 >/dev/null
```
Having some sort of test for this script is probably good.

But it seems kind of bad to have the test mutate the repo itself.

My suggestion is to make a copy, run versionize on the copy, and then run `mungedocs --verify` on the copy. My original (poorly worded!) complaint was that I had to run mungedocs after running versionize to get it to settle down. Sorry for the confusion!

nikhiljindal replied:
Yea that is what I tried first, but none of these scripts support an OUTPUT_DIR. Looking around, the tests for those scripts do the same thing as I have done with this one i.e copy the original files that will be mutated by the script, run the script to mutate the repo, do the comparison and then copy back the original files.
So I assumed its an accepted pattern :)

If you do want me to change the test, I think that will be a big enough change, which I will prefer to do in a separate PR. I can remove the test file from this PR for now, and send a separate PR with the tests redone? Or we can submit this as is and improve it later?

lavalamp replied:
> remove the test file from this PR for now, and send a separate PR with the tests redone

SGTM

nikhiljindal replied:
Removed

lavalamp commented on cmd/mungedocs/mungedocs.go:
```diff
@@ -46,10 +46,10 @@ Examples:
 	// TODO: allow selection from command line. (e.g., just check links in the examples directory.)
 	allMunges = []munge{
 		{"table-of-contents", updateTOC},
+		{"unversioned-warning", updateUnversionedWarning},
```
just curious: why the reordering?

nikhiljindal replied:
checkHeaderLines needs to run after this.
It adds blank lines above headings, but then this adds unversionedWarnings which messes up with those blank lines.

nikhiljindal commented:
Removed the test script and reverted the shippable change.
PTAL.

lavalamp commented on build/versionize-docs.sh:
```diff
@@ -0,0 +1,74 @@
+  SED=gsed
+fi
+if ! ($SED --version 2>&1 | grep -q GNU); then
+  echo "!!! GNU sed is required.  If on OS X, use 'brew install gnu-sed'."
```
exit 1?

nikhiljindal replied:
Done

lavalamp commented on build/versionize-docs.sh:
```diff
@@ -0,0 +1,74 @@
+
+if [ "$#" -lt 1 ]; then
+    echo "Usage: versionize-docs <release-version>"
+    exit
```
exit 1?

nikhiljindal replied:
Done

lavalamp commented:
Just tiny nits-- sorry I didn't notice earlier.

nikhiljindal commented:
Nice catch!
Thanks for pointing that out.
Have updated the code. PTAL.

lavalamp commented:
lgtm, thanks

nikhiljindal commented:
cc @vishh all green :)
